modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+                        F.raiseError[String](ConsumerGroupException(batch.consumerGroupIds))
+                      } else {
+                        F.pure(batch.consumerGroupIds.head)
+                      }
 
-                        Chunk
-                          .buffer(producerRecords)
-                          .traverse(
-                            KafkaProducer.produceRecord(keySerializer, valueSerializer, producer)
-                          )
-                          .map(_.sequence)
-                          .flatTap { _ =>
-                            F.delay {
-                              producer.sendOffsetsToTransaction(
-                                batch.offsets.asJava,
-                                groupId
-                              )
+                    consumerGroupId.flatMap { groupId =>
+                      records.flatTraverse { committable =>
+                        context
+                          .blockOn(blocker) {
+                            F.bracketCase(F.delay(producer.beginTransaction())) { _ =>
+                              committable.records
+                                .traverse(
+                                  KafkaProducer
+                                    .produceRecord(keySerializer, valueSerializer, producer)
+                                )
+                                .flatMap(_.sequence)
+                                .flatTap { _ =>
+                                  F.delay {
+                                    producer.sendOffsetsToTransaction(
+                                      batch.offsets.asJava,
+                                      groupId
+                                    )
+                                  }
+                                }
+                            } {
+                              case (_, ExitCase.Completed) =>
+                                F.delay(producer.commitTransaction())
+                              case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
+                                F.delay(producer.abortTransaction())
                             }
                           }
-                      } {
-                        case (_, ExitCase.Completed) =>
-                          F.delay(producer.commitTransaction())
-                        case (_, ExitCase.Canceled | ExitCase.Error(_)) =>
-                          F.delay(producer.abortTransaction())
                       }
-                    }.flatten
+                    }
                   }
-                }
-              }
 
-              override def toString: String =
-                "TransactionalKafkaProducer$" + System.identityHashCode(this)
+                override def toString: String =
+                  "TransactionalKafkaProducer$" + System.identityHashCode(this)
+              }
             }
           }
+        }
+      }
+
+    private def init(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
```

Comment:
You might want to make use of [`WithProducer`](https://github.com/ovotech/fs2-kafka/blob/master/modules/core/src/main/scala/fs2/kafka/internal/WithProducer.scala) in `TransactionalKafkaProducer`.

Reply:
I saw that, but I think the way it's set up will prevent manual setup / close management for `KafkaByteProducer`s that we'll need for the rebalancing case.

Reply:
> manual setup / close management

We'll need to have a `bracket` in there somewhere to ensure we `close` no matter what (e.g. on errors). But this can be done as a `Stream` with `Stream.bracket`.